### PR TITLE
[ Old Tracker Importer ] - Optimize query and entity caching [ DHIS2-12085 ]

### PR DESCRIPTION
**Pull request by lucaCambi77**
This issue aims to reduce the access to database objects during the old importer.
This required 2 steps:

- Add missing suppliers for prefetching and caching the entities
- Update old importers accordingly in order to use new suppliers

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/AbstractSupplier.java [hunk 1]
```diff
@@ -27,17 +27,75 @@
...
+                log.error( "Unable to read geometry for '" + entity + "': ", e );
+            }
+        }
+
+        return null;
+    }
+
+    public String getGeometryField( String field )
+    {
+        return SystemUtils.isH2( environment.getActiveProfiles() ) ? field : "ST_AsText( " + field + " )";
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/AbstractSupplier.java:99]: Can we try to avoid checking at runtime if this is a test or not?
We were trying to walk away from that as it is causing some performance issues in other parts of the system

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+         |
+         |
+         +---+ Organizational Unit (List)
+         |
+         +---+ User Access (ACL)
+         |
+         +---+ User Group Access (ACL)
+ * @formatter:on
+ *
+ * @author Luciano Fiandesio
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:102]: Identity theft is not a joke, Jim! Millions of families suffer every year!

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+
+                    programStage.setProgramStageDataElements( allProgramStageDataElements );
+                }
+            }
+        }
+    }
+
+    private Stream<DataElement> getOrEmpty( Set<DataElement> dataElementSet )
+    {
+        return Optional.ofNullable( dataElementSet )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:198]: In my opinion all these Optional.ofNullable are not adding any value to the code and they are making the flow harder to follow.
In here you are sure that a dataElementSet will never be null.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        final String sql = sqlSelect + " from program p "
+            + "LEFT JOIN categorycombo c on p.categorycomboid = c.categorycomboid "
+            + "LEFT JOIN trackedentitytype tet on p.trackedentitytypeid = tet.trackedentitytypeid "
+            + "LEFT JOIN programstage ps on p.programid = ps.programid "
+            + "LEFT JOIN program_organisationunits pou on p.programid = pou.programid "
+            + "LEFT JOIN organisationunit ou on pou.organisationunitid = ou.organisationunitid "
+            + "group by p.programid, tet.trackedentitytypeid, c.categorycomboid, ps.programstageid, ps.sort_order "
+            + "order by p.programid, ps.sort_order";
+
+        return Optional.ofNullable( jdbcTemplate.query( sql, ( ResultSet rs ) -> {
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:296]: Optional.ofNullable, same here.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+        dataElement.setCode( rs.getString( "de_code" ) );
+        return dataElement;
+    }
+
+    /**
+     * Check if the list of incoming Events contains one or more Program uid
+     * which is not in cache. Reload the entire program cache if a Program UID
+     * is not found
+     */
+    private boolean requiresCacheReload( List<Event> events, Map<String, Program> programMap )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:405]: Maybe not part of this PR but at some point it will be good to do something smarter here and reload the cache less frequently.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+import org.springframework.core.env.Environment;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+/**
+ * @author Luciano Fiandesio
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java:61]: Still not you :)

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+     * Instance by Program and Tracked Entity Instance
+     */
+    private void mapEventsToProgramInstanceByTei( ImportOptions importOptions, List<Event> events,
+        Map<String, ProgramInstance> programInstances, Map<String, Pair<TrackedEntityInstance, Boolean>> teiMap )
+    {
+        for ( Event event : events )
+        {
+            if ( !programInstances.containsKey( event.getUid() ) )
+            {
+                Pair<TrackedEntityInstance, Boolean> teiPair = teiMap.get( event.getUid() );
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java:125]: TeiMap and TeiPair are not really explicative.
What does the boolean represent here?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+
+    public Map<String, Program> get( ImportOptions importOptions, List<Event> eventList )
+    {
+        boolean requiresReload = false;
+
+        // Do not use cache if {@code skipCache} is true or if running as test
+
+        if ( importOptions.isSkipCache() )
+        {
+            programsCache.invalidateAll();
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:135]: Does it make sense to invalidate cache if this import is not using the cache?
What if there are 2 concurrent imports? Will this clear the cache used by the other import?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/WorkContextLoader.java [hunk 9]
```diff
@@ -180,16 +211,96 @@ private void initializeUser( ImportOptions importOptions )
         }
     }
 
-    /**
-     * Force Hibernate to pre-load all collections for the
-     * {@see UserCredentials} object and fetch the "isSuper()" data. This is
-     * required to avoid an Hibernate error later, when this object becomes
-     * detached from the Hibernate Session.
-     */
     private void initUserCredentials( UserCredentials userCredentials )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/WorkContextLoader.java:214]: Is the JavaDoc still relevant here?

**Review by enricocolasante — COMMENTED** [review]
Few punctual comments to have a look at.
Overall the PR is good, we will need to run performance tests against it as soon as possible to measure the impact of the changes

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/OrganisationUnitSupplier.java [hunk 10]
```diff
@@ -141,44 +107,47 @@ public OrganisationUnitSupplier( NamedParameterJdbcTemplate jdbcTemplate )
...
             while ( rs.next() )
             {
-                OrganisationUnit ou = mapFromResultSet( rs );
-
                 try
                 {
-                    for ( String event : orgUnitToEvent
+                    OrganisationUnit organisationUnit = mapFromResultSet( rs );
+
+                    for ( String event : orgUnitToEntity
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/OrganisationUnitSupplier.java:120]: Should event be entity here? maybe EntityIdentifier?

**Review by stian-sandvold — CHANGES_REQUESTED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+     */
+    private void mapExistingEventsToProgramInstances( ImportOptions importOptions, List<Event> events,
+        Map<String, ProgramInstance> programInstances )
+    {
+        // Collect all the Program Instances by event uid
+        final Map<String, ProgramInstance> programInstancesByEvent = getProgramInstanceByEvent( importOptions, events );
+
+        if ( !programInstancesByEvent.isEmpty() )
+        {
+            for ( Event event : events )
```
Comment (gnespolino) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java:158]: I think this for can be replaced by stream, collect and Collectors.toMap

**Review by gnespolino — COMMENTED** [review]
I didn't dig into the logic too much becase te PR is pretty big, just one coding style comment

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/AbstractSupplier.java (hunk 1 above)
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/AbstractSupplier.java:99]: I don't see how this is creating performance issues as is env variable check, in any case, the only way I can think of to improve it is to remove H2 tests or convert it to a not-h2 test

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java (hunk 3 above)
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:198]: This is a sonar suggestion

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java (hunk 4 above)
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:296]: again sonar issue, we can remove it ...

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java (hunk 7 above)
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java:125]: I have no idea what is this about, it is old code and I didn't tale the time to verify as this issue is about a complete different argument

**Comment by lucaCambi77** [comment]
@enricocolasante most of your comments refer to old code that I didn't take the time to verify. For caching if you have an alternative we can update here or in a separate issue. We can have a meeting

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java (hunk 2 above)
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java:102]: I didn't create this file, just minor updates for this issue

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 12]
```diff
@@ -530,7 +607,9 @@ private ImportSummary addEnrollment( Enrollment enrollment, ImportOptions import
 
         programInstanceService.addProgramInstance( programInstance, importOptions.getUser() );
 
-        importSummary = validateProgramInstance( program, programInstance, enrollment, importSummary );
+        sessionFactory.getCurrentSession().merge( daoTrackedEntityInstance );
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:610]: We add the program to the tei in this case. We use merge because it is not a session object. This way we prefetch objects not related to the session, avoid locks and still be able to update/persist them

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 13]
```diff
@@ -865,8 +937,8 @@ private ImportSummary updateEnrollment( Enrollment enrollment, ImportOptions imp
 
         if ( enrollment.getOrgUnit() != null )
         {
-            OrganisationUnit organisationUnit = getOrganisationUnit( importOptions.getIdSchemes(),
-                enrollment.getOrgUnit() );
+            OrganisationUnit organisationUnit = workContext.getOrganisationUnitMap().get( enrollment.getEnrollment() );
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:940]: see previous comment

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/importsummary/ImportSummary.java [hunk 14]
```diff
@@ -74,6 +74,18 @@
...
+    {
+        return updatedTei;
+    }
+
+    public void setUpdatedTei( String updatedTei )
+    {
+        this.updatedTei = updatedTei;
+    }
+
+    private String updatedTei;
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/importsummary/ImportSummary.java:87]: this is collecting all updated teis and do a single bulk update at the end

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 15]
```diff
@@ -562,18 +641,11 @@ private ImportSummary addEnrollment( Enrollment enrollment, ImportOptions import
         enrollment.setEnrollment( programInstance.getUid() );
         importSummary.getImportCount().incrementImported();
 
-        if ( handleEvents )
-        {
-            importSummary.setEvents( handleEvents( enrollment, programInstance, importOptions ) );
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:567]: not sure but this flag was always false. Events are persisted later on

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 16]
```diff
@@ -419,67 +538,25 @@ public ImportSummaries addEnrollments( List<Enrollment> enrollments, ImportOptio
...
                 : importOptions.getUser().getUsername());
 
-        if ( daoTrackedEntityInstance == null )
-        {
-            daoTrackedEntityInstance = getTrackedEntityInstance( enrollment.getTrackedEntityInstance(),
-                importOptions.getUser() );
-        }
+        TrackedEntityInstance daoTrackedEntityInstance = Optional.ofNullable( workContext.getTrackedEntityInstanceMap()
+            .get( enrollment.getTrackedEntityInstance() ) )
+            .orElseThrow( () -> new InvalidIdentifierReferenceException( "TrackedEntityInstance does not exist." ) );
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:553]: yes i agree but there are tests failing if you do that so I believe it is ok?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 17]
```diff
@@ -371,32 +385,137 @@ public ImportSummaries addEnrollments( List<Enrollment> enrollments, ImportOptio
...
+        importSummaries.addImportSummaries( importSummariesWithUpdatedTeis );
+
+        teiService.updateTrackedEntityInstanceLastUpdated( updatedTeis,
+            new Date() );
+
+        if ( !updatedTeis.isEmpty() )
+        {
+            String userName = User.username( user );
+
+            trackedEntityInstanceAuditService.addTrackedEntityInstanceAudit(
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:404]: previously we got the tei with getTrackedEntityInstance method that also was adding a read audit so I've just moved the logic out of that method but I am not sure we need that

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:535]: Which test is failing?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:592]: ok, I still think we should try to move any session-related work like this to the store, could we for example make a merge method in the store?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:567]: I thought they were persisted here? At least thats how it worked before,  the entire import is nested. Remember there is a difference between event from event programs and from tracker program; They are imported differently, since event program events are not nested within an enrollment for example.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/importsummary/ImportSummary.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (stian-sandvold) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/importsummary/ImportSummary.java:87]: ok, this is the the lastUpdated timestamp then?

**Comment by stian-sandvold** [comment]
I see the code coverage from sonarqube is not reporting here either.  Do you know how much of your changes are covered by tests? There is a lot of changes I think might be a bit risky, so I would like to spend some time to make sure we are not breaking anything.

**Comment by vilkg** [comment]
> I see the code coverage from sonarqube is not reporting here either. Do you know how much of your changes are covered by tests? There is a lot of changes I think might be a bit risky, so I would like to spend some time to make sure we are not breaking anything.
> 
> Especially, make sure that sync is working correctly, since its the main reason to keep this code at the moment.

Merging master in should help. Your fix for sonarqube was added 5 days ago, last merge was 6 days ago

**Review by stian-sandvold — APPROVED** [review]

**Comment by stian-sandvold** [comment]
Closed after conferring with Luca
